usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -5507,6 +6072,14 @@ CK_RV ep11tok_sign_init(STDLL_TokData_t * tokdata, SESSION * session,
         goto done;
     }
 
+    if (ep11tok_pkey_usage_ok(tokdata, session, key_obj, mech)) {
+        rc = sign_mgr_init(tokdata, session, ctx, mech, recover_mode, key);
```

Comment:
This comment applies to all places where ep11tok_pkey_usage_ok is called:

Because pkey_update_and_save (called within ep11tok_pkey_update which is called by ep11tok_pkey_usage_ok) unlocks and relocks the object, it might get concurrently updated by other threads at this place.

As far as I can see, there are references to object attributes (e.g. keyblob is pValue from CKA_IBM_OPAQUE of the object) used below this point. Those references may no longer be valid ! 

If you can ensure that the unlock/relock is ONLY performed when ep11tok_pkey_usage_ok returns TRUE, and thus the sign_mgr_init path is taken, then it might be OK.

Please check all places where ep11tok_pkey_usage_ok is called!

Reply:
You might think about moving h_opaque_2_blob below this point.... (Just an idea, don't know it this really works),

Reply:
I updated the return code handling of routine pkey_update_and_save to make sure it returns 0 only if all internal actions (unlock/relock, update attr) succeeded. There was in fact one case possible, where update attrs would fail, but the relocking succeeded afterwards, leading to a ret = 0. Regarding  h_opaque_2_blob: the flow is always: h_opaque_2_blob (obtains read lock) -> ep11tok_pkey_usage_ok, which may unlock/relock -> object_put to release the initial read-lock.

Reply:
That's not the point I was making. 

Once you unlock/relock the object, the objects attributes can be modified by other threads. This can affect ALL attributes of the object, not just the PKEY related attributes that you care of here. 

h_opaque_2_blob returns a pointer to the value of the CKA_IBM_OPAQUE attribute of the object. Once the object has been unlocked/relocked this pointer may no longer be valid, if the CKA_IBM_OPAQUE attribute has been changed by another thread. The code in the EP11 fallback DOES use that pointer, and thus may segfault if the pointer is no longer valid.

So you MUST ensure that you obtain the pointer to the key blob freshly, if you did unlock/relock the object, but then afterwards still go the EP11 fallback path.

Reply:
See places where ep11tok_pkey_check is called.